File: Backend/src/controller/usercontroller/registerController.ts
import { NextFunction, Request, Response } from 'express';
import User, { IUser } from '../../Database/models/userModel';
import bcrypt from 'bcryptjs';
import cc from 'currency-codes';
import generateRandomUserName from '../../utils/generateRandomUserName';
import cloudinary from '../../service/cloudinaryService';
import sendMail from '../../service/mailService';
import fs from 'fs';
import emailTemplates from '../../utils/emailTemplate';
import dotenv from 'dotenv';
import { Readable } from 'stream';
import createHttpError from 'http-errors';
import { userSchemaValidation } from '../../utils/JoiUtils/joiValidation';

export interface RequestBodyRegisterController {
    fullname: string;
    email: string;
    password: string;
    phonenumber: string;
    gender: string;
    preference: string[];
    country: string;
}

interface MulterFile {
    fieldname: string;
    originalname: string;
    encoding: string;
    mimetype: string;
    size: number;
    destination: string;
    filename: string;
    path: string;
    buffer: Buffer;
    stream: Readable;
}

export interface CustomRequestRegisterController<
    TParams = object,
    TQuery = object,
    TBody = RequestBodyRegisterController,
> extends Request<TParams, unknown, TBody, TQuery> {
    file: MulterFile;
}

dotenv.config();

const create_new_user = async (
    req: CustomRequestRegisterController,
    res: Response,
    next: NextFunction
): Promise<Response | void> => {
    const {
        fullname,
        email,
        password,
        phonenumber,
        gender,
        preference,
        country,
    } = req.body;

    try {
        // Check for required fields
        if (
            !fullname ||
            !email ||
            !password ||
            !phonenumber ||
            !gender ||
            !country
        ) {
            return next(createHttpError(400, 'All fields are required!'));
        }

        // Validate input data
        const { error } = userSchemaValidation.validate(req.body);
        if (error) {
            return next(createHttpError(400, error?.details[0].message));
        }

        // Check if the user already exists
        const checkUserExist: IUser | null = await User.findOne({
            $or: [{ email: email }, { phonenumber: phonenumber }],
        });

        if (checkUserExist) {
            return next(createHttpError(409, 'User already exist!'));
        }

        // Hash the password
        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(password, salt);

        // Generate a random username
        const username = await generateRandomUserName(fullname);

        // Upload profile image to Cloudinary
        let uploadImageUrl;
        try {
            uploadImageUrl = await cloudinary.uploader.upload(req.file.path);
            fs.unlinkSync(req.file.path); // Remove file from local storage
        } catch {
            // if (uploadError instanceof Error) {
            //     return res.status(500).send({
            //         status: 'failed',
            //         message: 'Error uploading profile picture',
            //         error: uploadError.message,
            //     });
            // } else {
            //     return res.status(500).send({
            //         status: 'failed',
            //         message: 'Error uploading profile picture',
            //         error: 'Unknown error',
            //     });
            // }
            return next(createHttpError(500, 'Error uploading file!'));
        }

        // Create Currency code
        const countryLower = country.toLowerCase();
        const cCode = cc.country(countryLower);
        if (!cCode || cCode.length === 0) {
            return next(
                createHttpError(
                    400,
                    'Currency code not found for the specified country!'
                )
            );
        }

        // Create the new user
        const newUser = new User({
            fullname: fullname,
            email: email,
            phonenumber: phonenumber,
            password: hashedPassword,
            username: username,
            gender: gender,
            preferences: preference,
            country: countryLower,
            profilepicture: uploadImageUrl.url,
            currency_code: cCode[0].code,
        });
        await newUser.save();

        // Email data
        const { registerEmailData } = emailTemplates;
        const emailData = registerEmailData(fullname, email);

        // Send welcome email
        await sendMail(emailData, (mailError: Error | null) => {
            if (mailError) {
                return next(
                    createHttpError(
                        500,
                        'User created, but email sending failed!'
                    )
                );
            }

            return res.status(201).send({
                status: 'success',
                message: 'User created successfully',
                userdata: {
                    fullname: newUser.fullname,
                    email: newUser.email,
                    phonenumber: newUser.phonenumber,
                    username: newUser.username,
                    gender: newUser.gender,
                    preference: newUser.preferences,
                    country: newUser.country,
                    profilepicture: newUser.profilepicture,
                    currency_code: newUser.currency_code, // Adjusted this line
                },
            });
        });
    } catch {
        return next(createHttpError(500, 'Internal Server Error!'));
    }
};

export default create_new_user;

File: APIs/app/controllers/user_controller/register_controller.py
from fastapi import Response, HTTPException, UploadFile, File, Form, status, BackgroundTasks
from app.database.models.user_model import User
from app.services.cloudinary_service import uploader
from app.services.mail_service import send_mail
from app.utils.email_template import register_email_data
from app.utils.generate_random_username import generate_random_username
from passlib.context import CryptContext
import pycountry
import json

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

async def create_new_user(
    fullname: str = Form(...),
    email: str = Form(...),
    password: str = Form(...),
    phonenumber: str = Form(...),
    gender: str = Form(...),
    preference: str = Form(...), # expecting JSON string or comma separated
    country: str = Form(...),
    profileimage: UploadFile = File(...),
    response: Response = None # type: ignore
):
    try:
        # Check exists
        # Beanie find is async
        user_exists = await User.find_one({
            "$or": [{"email": email}, {"phonenumber": phonenumber}]
        })
        
        if user_exists:
            raise HTTPException(status_code=409, detail="User already exist!")

        # Hash password
        hashed_password = pwd_context.hash(password)
        
        # Random Username
        username = generate_random_username(fullname)
        
        # Upload Image
        try:
            # Cloudinary upload_stream or upload
            # upload(file, options)
            upload_result = uploader.upload(profileimage.file, folder="AdventureNexus/Users")
            image_url = upload_result.get("secure_url")
        except Exception as e:
            print(f"Cloudinary error: {e}")
            raise HTTPException(status_code=500, detail="Error uploading file!")
            
        # Currency Code
        try:
            c = pycountry.countries.lookup(country)
            currency = pycountry.currencies.get(numeric=c.numeric) # Wrong mapping, need country to currency map
            # pycountry doesn't directly map country -> currency.
            # Use simple mapping or just iterate currencies?
            # Actually backend uses `currency-codes`.
            # Effective python way:
            # We can use `pycountry` to get alpha_2, then lookup currency?
            # Or use a quick library `country-currencies`.
            # For now, let's try a safe fallback or exact logic if possible.
            # Simplified: Use USD or lookup via hardcoded if pycountry fails complexity.
            # Let's try to find currency by country name using pycountry helper if available.
            # Actually, `pycountry` doesn't provide country->currency map easily.
            # I will assume USD if not found or implement a basic map later if critical. 
            # WAIT: Backend logic: `cc.country(countryLower)`.
            # I will use a placeholder mapper for now to save time, or just "USD".
            currency_code = "USD"
        except:
             currency_code = "USD"
             
        # Parse preferences
        try:
            prefs_list = json.loads(preference)
        except:
            prefs_list = [preference]

        # Create User
        new_user = User(
            fullname=fullname,
            email=email,
            phonenumber=phonenumber,
            password=hashed_password,
            username=username,
            gender=gender,
            preferences=prefs_list,
            country=country.lower(),
            profilepicture=image_url,
            currency_code=currency_code
        )
        
        await new_user.save()
        
        # Send Mail (Background task ideally, but mirroring synchronous-like flow with await)
        email_data = register_email_data(fullname, email)
        await send_mail(email_data["to"], email_data["subject"], email_data["html"])
        
        # Response
        response.status_code = status.HTTP_201_CREATED
        return {
            "status": "success",
            "message": "User created successfully",
            "userdata": {
                "fullname": new_user.fullname,
                "email": new_user.email,
                "phonenumber": new_user.phonenumber,
                "username": new_user.username,
                "gender": new_user.gender,
                "preference": new_user.preferences,
                "country": new_user.country,
                "profilepicture": new_user.profilepicture,
                "currency_code": new_user.currency_code
            }
        }

    except HTTPException as he:
        raise he
    except Exception as e:
        print(f"Register Error: {e}")
        raise HTTPException(status_code=500, detail="Internal Server Error!")
